### DISCUSSION TITLE
FutureSDR as an app sub-component

### DISCUSSION THREAD

**Discussion by etschneider** [FutureSDR/FutureSDR#76] [Ideas]
@bastibl, this is essentially a follow-up to your reply from issue #70, regarding block initialization and flowgraph reconfiguration via serializable builders. The scope of my thoughts are pretty broad and future oriented, so I figured a discussion topic might be best.

I'm hoping to be able to use FutureSDR within another pretty heavy app. There are a few things that don't quite seem like they fit quite right. And some plans, as I understand them, even less so. I will have a suggested resolution at the end, and I mean all of this as no more than a use case example for consideration.

My app already has: a web API, a configuration system, and a logging system. So, as a user of FutureSDR as a sub-component, I'd like:

- my app to provide the web server for control port instead of the reverse
- configuration data to be queried from the apps config
- let my logging system assign logging levels and such

These are all pretty easily dealt with. E,g, disable control port, use the `flowgraphHandle` myself. Config and logging are pretty simple, so no real biggie.

But it seems to be indicative of an expectation that FutureSDR will be the _provider_ of such higher level services. That is what doesn't feel right.

Your serializable builders goal is similar in my mind. I really do like the idea of a fully serializable flowgraph. It would be great, and has a lot of uses.

But... That serialization interface defines the _maximum_ flexibility of the system. If it is not in the builder definition, you can't do it. A "config jail", so to speak. And I can see those builders getting pretty outrageous, especially for something as abstract as the Soapy interface.

So I think the right approach is to provide two different levels of interfaces. One that provides as low level of control as is reasonable, and another that builds on top of that low-level interface that provides the "batteries included" experience that makes current FutureSDR demos so impressive. Plumbing and porcelain.

The serializable builder for Soapy begs a simple solution. Remove `dev` from the builder, but keep it in the constructor. If I need the flexibility for what I do with a dev, I can just make an instance without the builder. That also relieves the burden of maintaining a builder setting for every possible whim. If I really need the reconfiguration capability that only a builder will provide, then I can make a one-off builder modification for my special case.

Whew, all done. Seem sane?

**Comment by bastibl** [comment]
While it may seem like it, I don't see FutuerSDR as an app but more like a library, it's just that at this development state, it's easier to just hack something together in the FutureSDR repo, since it is likely that something has to be changed in the runtime. But regarding your issues:

You can configure logging and FutureSDR will happily pick it up. Just try adding `env_logger` at the start and FutureSDR will use it. The dummy logger is only instantiated, if there is no logger available when starting the runtime. This should make quick hacks easier and more concise. There is also an example with a custom logger: https://github.com/FutureSDR/FutureSDR/blob/main/examples/logging/logging.rs. It just has to be compatible with the `log` crate.

While the current examples mainly just use the built-in control port, there is quite a lot flexibility. One can
- use control port
- add custom routes to it: https://github.com/FutureSDR/FutureSDR/tree/main/examples/custom-routes
- use a different frontend: https://github.com/FutureSDR/FutureSDR/blob/main/src/runtime/config.rs#L97
- use the flowgraph handle in a custom web server

I was quite happy with this, since it provides a lot flexibility without causing complexity in the runtime implementation. What else are you missing? Quick hacks get control port for free; more involved apps have full flexibility.

I'm also not really happy with the config at the moment, but I also didn't have a good idea yet. As a quick hack, your app might just set environment variables and the config will pick them up. Another better, but still sub-optimal, solution is to allow setting the global config. This could be implemented trivially. What would be ideal was a config that is associated with the runtime. But then, one would have to instantiate it first and create flowgraphs and blocks from the runtime. Otherwise, blocks have no way to access the config. E.g.:

```rust
let mut rt = Runtime::with_config(config);
let flowgraph = rt.new_flowgraph();
let block = rt.create_block(block_builder);
```

Just some rough API, but I think something like this is needed to allow accessing the config in the flowgraph/block constructors.

Regarding serializable builders, I'm not sure I understand the problem. When one wants to allow, for example, setting up a flowgraph remotely, through a web interface, or generally an external process (I want to decouple the GUI from the runtime), then this is just needed. But with Rust, I don't see an issue or that it would cause any limitations. Some blocks may just implement serialization and this subset of blocks can be used remotely. Other blocks do not have to and will just be used like they are now. Same spirit as with the Pmt. If some types implement serializable `Any` that's great, but nobody has to. (For the Soapy example, one would, in essence, just need another builder for the dev.)

**Reply by etschneider** [reply]
Just to be clear, this is just me trying to visualize pain points in the future, as the ecosystem grows. I have no immediate problems, I'm just putting out ideas. I get that you are busy, no rush, no pressure. I've just got a little time between projects and wanted to dig into FutureSDR when I had the opportunity.

Regarding the control_port/config/log architecture (without too much thought about the specifics). I'd separate runtime into two structs, like `runtime_core` and `runtime`.  `runtime_core` would be the current implementation, minus control_port and config. `runtime` would then encapsulate `runtime_core` and add the control port server and default config implementation as so on. Typical users still see `runtime` as it is now, but under the hood is the core, which doesn't know anything about the higher level functions.

Mostly a design preference I guess. I like my design units lean and loosely coupled, perhaps to an unreasonable degree.

**Reply by etschneider** [reply]
Regarding serializable builders, if the builders require serialization traits (e.g. because FutureSDR needs them for flowgraph serialization), then things like `dev` can't be used directly because it does not support serialization.

I prefer that the use of builders is not _required_ in the future. For instance because the flowgraph requires builders for all blocks to support reconfig/remote/etc functionality. Just thoughts to consider for your future design choices.
